### ISSUE TITLE
Listener.onPointerPanZoomEnd does not always fire with mac trackpad

### ISSUE THREAD

**Issue by bramp** [flutter/flutter#143179]
### Steps to reproduce

On a MacBook Pro (Apple M2 Pro - MacOS 14.2.1) trackpad, two finger pinch to zoom.

### Expected results

This gets printed

```
flutter: onPointerPanZoomStart _TransformedPointerPanZoomStartEvent#f0774(position: Offset(414.2, 262.3))
flutter: onPointerPanZoomEnd _TransformedPointerPanZoomEndEvent#052c2(position: Offset(414.2, 262.3))
```

### Screenshots or Video

<details open>
<summary>Screenshots / Video demonstration</summary>

https://github.com/flutter/flutter/assets/160627/40925c1f-5bb9-40ae-bd4a-7acbdc8f15ac

</details>

### Actual results

Sometimes the `onPointerPanZoomEnd` does not get fired at the end of the pinch (i.e when fingers have been taken off trackpad). It does get fired on a subsequent pan/zoom.

The video shows this, at the beginning the pinch zoom is working fine, then at 0:12 it stops, but a better example is at 0:16

```
# 0:16 in video (start to two finger pinch and release)
flutter: onPointerPanZoomStart _TransformedPointerPanZoomStartEvent#2e3ea(position: Offset(414.4, 261.6))

# stuck here. I can put my finger back on the touch pad, and you can clearly see my cursor moving.

# 0:20 in video (after I tapped)
flutter: onPointerDown _TransformedPointerDownEvent#e8991(position: Offset(434.8, 243.4))
flutter: onPointerUp _TransformedPointerUpEvent#2511d(position: Offset(434.8, 243.4))

# 0:22 in video - I started another zoom here, which triggered the onPointerPanZoomEnd.
# typically the start/end positions are the same. But not in this case (expected Offset(414.4, 261.6))
flutter: onPointerPanZoomEnd _TransformedPointerPanZoomEndEvent#50b58(position: Offset(415.1, 407.2))
```

This does not happen every pinch/zoom but I can make it happen after a few tries.  I'm really not sure what I'm doing with my two fingers to trigger this.

### Code sample

<details open><summary>Code sample</summary>

Tweaked example app from a [similar bug](https://github.com/flutter/flutter/issues/112103)

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      title: 'MyApp Demo',
      home: Scaffold(
        body: Center(
          child: SingleChildScrollView(
            child: MainWidget(),
          ),
        ),
      ),
      debugShowCheckedModeBanner: false,
    );
  }
}

class MainWidget extends StatefulWidget {
  const MainWidget({Key? key}) : super(key: key);

  @override
  State<MainWidget> createState() => _MainWidgetState();
}

class _MainWidgetState extends State<MainWidget> {
  double scale = 1.0;

  double scaleStart = 1.0;

  @override
  Widget build(BuildContext context) {
    return Listener(
      onPointerPanZoomStart: (event) {
        print("onPointerPanZoomStart $event");
        scaleStart = scale;
      },
      onPointerPanZoomUpdate: (event) {
        //print("onPointerPanZoomUpdate $event");
        setState(() {
          scale = scaleStart * event.scale;
        });
      },
      onPointerPanZoomEnd: (event) {
        print("onPointerPanZoomEnd $event");
        scaleStart = scale;
      },
      onPointerCancel: (event) => print("onPointerCancel $event"),
      onPointerDown: (event) => print("onPointerDown $event"),
      onPointerUp: (event) => print("onPointerUp $event"),
      child: Container(
        // This will ensure that the entire window space is covered by the
        // listener.
        width: MediaQuery.of(context).size.width,
        height: MediaQuery.of(context).size.height,
        // The color ensure that the hit test succeeds.
        decoration: BoxDecoration(
          color: Colors.green.withOpacity(0.1),
          border: Border.all(color: Colors.black, width: 3),
        ),
        child: Stack(
          alignment: Alignment.center,
          children: [
            Transform.scale(
              scale: scale,
              child: Center(
                child: Container(
                  width: 300,
                  height: 300,
                  decoration: BoxDecoration(
                    border: Border.all(color: Colors.black, width: 3),
                    gradient: const LinearGradient(
                      begin: Alignment.topLeft,
                      end: Alignment.bottomRight,
                      colors: [
                        Colors.red,
                        Colors.green,
                        Colors.blue,
                      ],
                    ),
                  ),
                ),
              ),
            ),
            Material(
              color: Colors.transparent,
              child: InkWell(
                onTap: () {
                  setState(() {
                    scale = 1.0;
                  });
                },
                child: Container(
                    decoration: BoxDecoration(
                      border: Border.all(color: Colors.black, width: 1),
                      color: Colors.white.withOpacity(0.5),
                      borderRadius: BorderRadius.circular(4),
                    ),
                    padding: const EdgeInsets.all(4),
                    child: Text('Scale: ${scale.toStringAsPrecision(3)}')),
              ),
            ),
          ],
        ),
      ),
    );
  }
}
```

</details>

### Logs

<details open><summary>Logs</summary>

```console

flutter: onPointerPanZoomStart _TransformedPointerPanZoomStartEvent#f0774(position: Offset(414.2, 262.3))
flutter: onPointerPanZoomEnd _TransformedPointerPanZoomEndEvent#052c2(position: Offset(414.2, 262.3))

flutter: onPointerPanZoomStart _TransformedPointerPanZoomStartEvent#dc0ad(position: Offset(414.2, 262.3))
flutter: onPointerPanZoomEnd _TransformedPointerPanZoomEndEvent#13d42(position: Offset(414.2, 262.3))

flutter: onPointerPanZoomStart _TransformedPointerPanZoomStartEvent#587f3(position: Offset(414.2, 262.3))
flutter: onPointerPanZoomEnd _TransformedPointerPanZoomEndEvent#948a7(position: Offset(414.2, 262.3))

flutter: onPointerPanZoomStart _TransformedPointerPanZoomStartEvent#17dcb(position: Offset(414.2, 262.3))
flutter: onPointerPanZoomEnd _TransformedPointerPanZoomEndEvent#bd994(position: Offset(414.2, 262.3))
flutter: onPointerPanZoomStart _TransformedPointerPanZoomStartEvent#711b8(position: Offset(414.2, 262.3))
flutter: onPointerPanZoomEnd _TransformedPointerPanZoomEndEvent#5344e(position: Offset(414.2, 262.3))
flutter: onPointerPanZoomStart _TransformedPointerPanZoomStartEvent#e84b7(position: Offset(414.2, 262.3))
flutter: onPointerPanZoomEnd _TransformedPointerPanZoomEndEvent#9f048(position: Offset(414.2, 262.3))
flutter: onPointerPanZoomStart _TransformedPointerPanZoomStartEvent#e3031(position: Offset(414.2, 262.3))
flutter: onPointerPanZoomEnd _TransformedPointerPanZoomEndEvent#59e7d(position: Offset(414.2, 262.3))
flutter: onPointerPanZoomStart _TransformedPointerPanZoomStartEvent#3dea6(position: Offset(414.2, 262.3))
flutter: onPointerPanZoomEnd _TransformedPointerPanZoomEndEvent#a389c(position: Offset(414.2, 262.3))
flutter: onPointerPanZoomStart _TransformedPointerPanZoomStartEvent#fc1f1(position: Offset(414.2, 262.3))
flutter: onPointerPanZoomEnd _TransformedPointerPanZoomEndEvent#d0762(position: Offset(414.2, 262.3))
flutter: onPointerPanZoomStart _TransformedPointerPanZoomStartEvent#b2d03(position: Offset(414.2, 262.3))
flutter: onPointerPanZoomEnd _TransformedPointerPanZoomEndEvent#0076c(position: Offset(414.2, 262.3))
flutter: onPointerPanZoomStart _TransformedPointerPanZoomStartEvent#17f44(position: Offset(414.4, 261.6))
flutter: onPointerPanZoomEnd _TransformedPointerPanZoomEndEvent#9c57e(position: Offset(414.4, 261.6))
flutter: onPointerPanZoomStart _TransformedPointerPanZoomStartEvent#8e3a6(position: Offset(414.4, 261.6))
flutter: onPointerPanZoomEnd _TransformedPointerPanZoomEndEvent#cfb5a(position: Offset(414.4, 261.6))
flutter: onPointerPanZoomStart _TransformedPointerPanZoomStartEvent#5f71d(position: Offset(414.4, 261.6))
flutter: onPointerPanZoomEnd _TransformedPointerPanZoomEndEvent#f399b(position: Offset(414.4, 261.6))

# 0:12 in the video
flutter: onPointerPanZoomStart _TransformedPointerPanZoomStartEvent#531e5(position: Offset(414.4, 261.6))
# got stuck here, but I too quickly tapped for it to be noticable.

flutter: onPointerDown _TransformedPointerDownEvent#31ded(position: Offset(414.4, 261.6))
flutter: onPointerUp _TransformedPointerUpEvent#871ef(position: Offset(414.4, 261.6))
flutter: onPointerPanZoomEnd _TransformedPointerPanZoomEndEvent#cacba(position: Offset(414.4, 261.6))

# 0:16 in video (start to two finger pinch and release)
flutter: onPointerPanZoomStart _TransformedPointerPanZoomStartEvent#2e3ea(position: Offset(414.4, 261.6))

# stuck here. Only one finger on the touch bad, and you can clearly see my cursor moving.

# 0:20 in video (after I tapped)
flutter: onPointerDown _TransformedPointerDownEvent#e8991(position: Offset(434.8, 243.4))
flutter: onPointerUp _TransformedPointerUpEvent#2511d(position: Offset(434.8, 243.4))

# 0:22 in video - I started another zoom here, which triggered the onPointerPanZoomEnd.
# typically the start/end positions are the same. But not in this case (expected Offset(414.4, 261.6))
flutter: onPointerPanZoomEnd _TransformedPointerPanZoomEndEvent#50b58(position: Offset(415.1, 407.2))

# next zoom works, but offsets mismatch.
flutter: onPointerPanZoomStart _TransformedPointerPanZoomStartEvent#aed33(position: Offset(415.1, 407.2))
flutter: onPointerPanZoomEnd _TransformedPointerPanZoomEndEvent#e78ec(position: Offset(252.3, 226.0))
```

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
❯ flutter doctor -v
[✓] Flutter (Channel stable, 3.16.9, on macOS 14.2.1 23C71 darwin-arm64, locale en-GB)
    • Flutter version 3.16.9 on channel stable at /opt/homebrew/Caskroom/flutter/3.16.0/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 41456452f2 (2 weeks ago), 2024-01-25 10:06:23 -0800
    • Engine revision f40e976bed
    • Dart version 3.2.6
    • DevTools version 2.28.5

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/bramp/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b829.9-10027231)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.2)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 15C500b
    • CocoaPods version 1.15.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2022.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b829.9-10027231)

[✓] VS Code (version 1.86.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.83.20240201

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-arm64   • macOS 14.2.1 23C71 darwin-arm64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 121.0.6167.160

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

</details>

**Comment by bramp** [comment]
I found this in a larger app using a InteractiveViewer, wrapped in a GestureDetector. When the onPointerPanZoomEnd does not fire, the GestureDetector.onTap stops working, causing functionality in the app to stop working. This minimum example seems to reproduce the same behaviour.

I'm happy to help debug more, I just don't know where to go from here. I've also not tested this on other devices.

**Comment by moffatman** [comment]
I fixed this recently... try master channel, it should be ok there

**Comment by bramp** [comment]
Doh, I didn't not find that fix / related bugs before I reported this issue.

But thanks, this certainly seems to have fixed my issue!

However, in my test cases above, I was able to trigger two onPointerPanZoomStart (in quick succession) with only one End after some aggressive zoom/panning.

```
flutter: onPointerPanZoomStart _TransformedPointerPanZoomStartEvent#a17a6(position: Offset(367.2, 357.8))
flutter: onPointerPanZoomEnd _TransformedPointerPanZoomEndEvent#83e0a(position: Offset(367.2, 357.8))

flutter: onPointerPanZoomStart _TransformedPointerPanZoomStartEvent#633a7(position: Offset(367.2, 357.1))
flutter: onPointerPanZoomStart _TransformedPointerPanZoomStartEvent#d5348(position: Offset(398.8, 259.5))
flutter: onPointerPanZoomEnd _TransformedPointerPanZoomEndEvent#82157(position: Offset(398.8, 259.5))

flutter: onPointerPanZoomStart _TransformedPointerPanZoomStartEvent#64bee(position: Offset(398.8, 259.5))
flutter: onPointerPanZoomEnd _TransformedPointerPanZoomEndEvent#6cef9(position: Offset(398.8, 259.5))
```

but I was only able to trigger that once.  I don't know if that requires some investigation? Otherwise feel free to resolve this issue. Thanks.

**Comment by moffatman** [comment]
Since the next start has a different x,y position. It looks like there is still some cause of a missing end event.

**Comment by moffatman** [comment]
Can you do your test again with timestamps? Thx

**Comment by bramp** [comment]
Sure, I added

```dart
String now() {
  DateTime now = DateTime.now();
  return '${now.hour}:${now.minute}:${now.second}.${now.millisecond}';
}

...
print("${now()} onPointerPanZoomStart $event");
```

and I just spent a good few minutes trying to repo, but I can't :(.  Gah!   Maybe I wasn't correctly using master earlier, or it's just a fluke.  I'll report back if I can repo this.

**Comment by dam-ease** [comment]
Thanks for your input @moffatman.
Hello @bramp.
I tested on the latest `master` channel as well and can see this issue no longer reproduces there (see logs below). Can you confirm this as well?

<details><summary>Logs</summary>
<p>

```
flutter: 11:23:5.238 onPointerPanZoomStart _TransformedPointerPanZoomStartEvent#62919(position: Offset(315.1, 148.8))
flutter: onPointerPanZoomEnd _TransformedPointerPanZoomEndEvent#51d75(position: Offset(315.1, 148.8))
flutter: 11:23:7.265 onPointerPanZoomStart _TransformedPointerPanZoomStartEvent#f1959(position: Offset(300.3, 141.7))
flutter: onPointerPanZoomEnd _TransformedPointerPanZoomEndEvent#79e65(position: Offset(300.3, 141.7))
flutter: 11:23:9.263 onPointerPanZoomStart _TransformedPointerPanZoomStartEvent#f0acf(position: Offset(300.3, 141.7))
flutter: onPointerPanZoomEnd _TransformedPointerPanZoomEndEvent#8c589(position: Offset(300.3, 141.7))
flutter: 11:23:9.667 onPointerPanZoomStart _TransformedPointerPanZoomStartEvent#107f6(position: Offset(300.3, 141.7))
flutter: onPointerPanZoomEnd _TransformedPointerPanZoomEndEvent#5bef7(position: Offset(300.3, 141.7))
flutter: 11:23:12.715 onPointerPanZoomStart _TransformedPointerPanZoomStartEvent#17f10(position: Offset(300.3, 141.7))
flutter: onPointerPanZoomEnd _TransformedPointerPanZoomEndEvent#290dc(position: Offset(300.3, 141.7))
flutter: 11:23:12.865 onPointerPanZoomStart _TransformedPointerPanZoomStartEvent#db174(position: Offset(300.3, 141.7))
flutter: onPointerPanZoomEnd _TransformedPointerPanZoomEndEvent#3d075(position: Offset(300.3, 141.7))
``` 

</p>
</details> 

<details><summary>master flutter doctor -v</summary>
<p>

```
[!] Flutter (Channel master, 3.20.0-4.0.pre.62, on macOS 14.2.1 23C71 darwin-arm64, locale en-NG)
    • Flutter version 3.20.0-4.0.pre.62 on channel master at /Users/damilolaalimi/fvm/versions/master
    ! Warning: `dart` on your path resolves to /opt/homebrew/Cellar/dart/3.1.5/libexec/bin/dart, which is not inside your current Flutter SDK checkout at /Users/damilolaalimi/fvm/versions/master. Consider adding /Users/damilolaalimi/fvm/versions/master/bin to the front of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 950e0025c4 (4 hours ago), 2024-02-09 00:38:30 -0500
    • Engine revision f6dddb257a
    • Dart version 3.4.0 (build 3.4.0-116.0.dev)
    • DevTools version 2.32.0
    • If those were intentional, you can disregard the above warnings; however it is recommended to use "git" directly to perform update checks and upgrades.

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/damilolaalimi/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0
    • ANDROID_HOME = /Users/damilolaalimi/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.2)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 15C500b
    • CocoaPods version 1.14.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2022.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)

[!] Android Studio (version unknown)
    • Android Studio at /Users/damilolaalimi/Downloads/Android Studio Preview.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    ✗ Unable to determine Android Studio version.
    • Java version OpenJDK Runtime Environment (build 17.0.7+0-17.0.7b1000.6-10550314)

[✓] VS Code (version 1.86.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.78.0

[✓] VS Code (version 1.83.1)
    • VS Code at /Users/damilolaalimi/Downloads/Visual Studio Code.app/Contents
    • Flutter extension version 3.78.0

[✓] Connected device (5 available)
    • sdk gphone64 arm64 (mobile)     • emulator-5554             • android-arm64  • Android 14 (API 34) (emulator)
    • Damilola’s iPhone (mobile)      • 00008110-001964480AE1801E • ios            • iOS 17.1.1 21B91
    • macOS (desktop)                 • macos                     • darwin-arm64   • macOS 14.2.1 23C71 darwin-arm64
    • Mac Designed for iPad (desktop) • mac-designed-for-ipad     • darwin         • macOS 14.2.1 23C71 darwin-arm64
    • Chrome (web)                    • chrome                    • web-javascript • Google Chrome 121.0.6167.160

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 2 categories.
``` 

</p>
</details>

**Comment by bramp** [comment]
Ok I tried again now, and again I can not reproduce it. So either my [original test](https://github.com/flutter/flutter/issues/143179#issuecomment-1934963103) on master was bad, or I hit a different edge case that I can't reproduce.

For the moment, everything seems good. Thanks all!
